pennylane/_qubit_device.py

```diff
@@ -161,6 +174,14 @@ def execute(self, circuit, **kwargs):
         Returns:
             array[float]: measured value(s)
         """
+        if self._caching:
+            try:  # TODO: Remove try/except when circuit is always QuantumTape
+                circuit_hash = circuit.graph.hash
```

Comment:
Why would this fail in non-tape mode?

Reply:
@trbromley I had no idea this behaviour had changed! It definitely wasn't intentional 😆 But it would be good to know why it has changed.

Further, is it a bug, or an 'improvement'?

Reply:
@josh146, it all comes down to `serialize()`:

Consider
```python
import pennylane as qml

dev = qml.device("default.qubit", wires=2, shots=100, analytic=False)

@qml.qnode(dev)
def f(w):
    qml.Rot(*w, wires=0)
    qml.RX(0.3, wires=1)
    qml.CNOT(wires=[0, 1])
    return qml.expval(qml.PauliZ(0))

f([0.3, 0.4, 0.5])
```

**Tape mode:**

```pycon
>>> f.qtape.graph.serialize()
'Rot!0.3!!0.4!!0.5![0]RX!0.3![1]CNOT[0, 1]|||PauliZ[0]'
```

**Non-tape mode:**

```pycon
'Rot!V0!!V1!!V2![0]RX!0.3![1]CNOT[0, 1]|||PauliZ[0]'
```

It all comes down to replacing of `Variables`! :exploding_head: 

**What about TF/Torch in tape mode?!:**

TF
```pycon
>>> f.qtape.graph.serialize()
'Rot!tf.Tensor(0.3, shape=(), dtype=float32)!!tf.Tensor(0.4, shape=(), dtype=float32)!!tf.Tensor(0.5, shape=(), dtype=float32)![0]RX!0.3![1]CNOT[0, 1]|||PauliZ[0]'
```


Torch
```pycon
>>> f.qtape.graph.serialize()
'Rot!tensor(0.3000)!!tensor(0.4000)!!tensor(0.5000)![0]RX!0.3![1]CNOT[0, 1]|||PauliZ[0]'
```

This seems ok, we are serializing the tensors rather than trying to hash them.

Reply:
Got it - so essentially, the device side caching is more 'robust' in tape mode than it is in non-tape mode, via a quirk of implementation